deepspeed/runtime/pipe/engine.py

```diff
@@ -1305,6 +1295,18 @@ def _allocate_buffers(self, shapes_and_dtypes, requires_grad=False, num_buffers=
             buffers.append(buffer)
         return buffers
 
+    def _allocate_or_extend_buffers(self, idx, shape, dtype):
+        numel = reduce(mul, shape) if len(shape) > 0 else 1
+        if len(self._grad_layer_buf) <= idx or self._grad_layer_buf[idx].numel() < numel:
+            new_buf = self._allocate_buffer(shape, dtype=dtype, num_buffers=1)[0]
+            if len(self._grad_layer_buf) <= idx:
+                self._grad_layer_buf.append(new_buf)
```

Comment:
why could len(self._grad_layer_buf) be smaller then idx?  Dynamic shape inputs shouldn't change the number of gradients, right?

Reply:
The original code also extended the buffer for gradients (`_allocate_buffers`). I added `_allocate_or_extend_buffers` to pass shape and dtype one by one, instead of having a loop inside.
I found that `_allocate_buffers` is now unused. Let me just delete it.